### DISCUSSION TITLE
Docker EE Contraints Dilemma

### DISCUSSION THREAD

**Discussion by rfults** [BretFisher/ama#186] [Q&A]
I have several 35-40 node Swarm clusters running UCP with anywhere from 120-400 services with multiple replicas.  I have one ridiculously critical service that needs to be isolated to a couple dedicated workers in each cluster AND it needs to be the only service on those workers.  How can I go about setting this up without having to add labels to all of my other services in order to prevent them from deploying to these nodes whilst allowing this one service to deploy on them?  Docker support suggested using collections however gave very little guidance around configuring them for this sort of isolation and the docs really don't seem to have a clear answer around this either.

**Comment by guillaumemaka** [comment]
Hi @rfults,

You can add labels to the worker node, and constrained the service to be place on the worker with the defined labels.

https://success.docker.com/article/using-contraints-and-labels-to-control-the-placement-of-containers

Hope that's help.

**Comment by rfults** [comment]
Unfortunately that doesn't resolve this part of the problem: 

> without having to add labels to all of my other services

**Comment by BretFisher** [comment]
I don't know of a way to do that without updating existing services. Not that the label updates won't redeploy the containers unless they are on the affected nodes.

**Comment by mistresseve666** [comment]
@rfults Docker EE Collections are a fancy wrapper around labels. See more here:

https://docs.docker.com/v17.09/datacenter/ucp/2.2/guides/access-control/manage-access-with-collections/

TL;DR: Add a label `com.docker.ucp.access.label="/Shared/myapp"` and the container will be started on a node under the `/Shared/myapp` collection.

**Comment by rfults** [comment]
I guess I just have to suck it up and add labels to everything. :(

**Comment by rfults** [comment]
Wrapping this up, we got the developer to fix the service so it could actually be scaled across multiple instances without exploding now and ended up not using the labels.  We've also since migrated to k8s. :D

**Comment by BretFisher** [comment]
Reopening so others can learn, thanks.
